vllm/model_executor/models/ultravox.py

```diff
@@ -0,0 +1,441 @@
+        assert self.multi_modal_config
+
+        if config.audio_model_id is not None:
+            self.audio_tower = ModifiedWhisperEncoder.from_pretrained(
+                config.audio_model_id)
+        else:
+            self.audio_tower = ModifiedWhisperEncoder(config.audio_config)
+        self.multi_modal_projector = UltravoxProjector(config)
+        self.language_model = init_vllm_registered_model(
+            config.text_config, cache_config, quant_config)
+
+    def _audio_features_to_embeddings(self, input_features: torch.Tensor,
+                                      dtype: torch.dtype) -> torch.Tensor:
+        audio_input = input_features.to(self.audio_tower.dtype)
+        audio_features = self.audio_tower(audio_input).last_hidden_state
+        audio_features = audio_features.to(self.audio_tower.dtype)
+        audio_embeddings = self.multi_modal_projector(audio_features).to(dtype)
+        return audio_embeddings
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+        intermediate_tensors: Optional[torch.Tensor],
+        *,
+        input_features: Optional[BatchedTensors] = None,
+    ) -> SamplerOutput:
+        """Run forward pass for Ultravox
+
+        One key thing to understand is the `input_ids` already accounts for the
+        positions of the to-be-inserted audio embeddings. The to-be-inserted
+        audio has a size that is essentially 6.25 tokens per second of audio.
+
+        This way, the `positions` and `attn_metadata` are consistent
+        with the `input_ids`.
+
+        Args:
+            input_features: A batch of audio inputs, [1, 80, M].
+        """
+        if input_features is not None:
+            inputs_embeds = self.language_model.model.get_input_embeddings(
+                input_ids)
+            if isinstance(input_features, list):
+                # TODO: Batch these through the encoder/projector instead of
+                # serializing them.
+                audio_embeddings = [
+                    self._audio_features_to_embeddings(
+                        single_features.unsqueeze(0),
+                        inputs_embeds.dtype).squeeze(0)
+                    for single_features in input_features
+                ]
+            elif isinstance(input_features, torch.Tensor):
+                audio_embeddings = self._audio_features_to_embeddings(
+                    input_features, inputs_embeds.dtype)
+            else:
+                raise ValueError(
+                    "The input audio features should be a tensor or a list "
+                    f"of tensors, not {type(input_features)}")
+
+            merge_multimodal_embeddings(input_ids, inputs_embeds,
+                                        audio_embeddings,
+                                        _AUDIO_PLACEHOLDER_TOKEN)
+            input_ids = None
```

Comment:
Could you adopt the same pattern as we have in other image models for this portion? i.e, 

        image_input = self._parse_and_validate_image_input(**kwargs)

        if image_input is not None:
            vision_embeddings = self._process_image_input(image_input)
            inputs_embeds = self.language_model.model.get_input_embeddings(
                input_ids)

            inputs_embeds = merge_multimodal_embeddings(
                input_ids, inputs_embeds, vision_embeddings,
                self.config.image_token_index)

            input_ids = None
        else:
            inputs_embeds = None

It would be great if we can keep the inference pipeline pattern as consistent as possible across the codebase. Let me know if you have any strong opinion about this!

Reply:
No problem -- I tried to align these but let me know if I missed something.